src/solver/mha/mha_ck_fa_v2_solver_forward.cpp

```diff
@@ -0,0 +1,265 @@
+    fmha_args.nhead_stride_bias = 0;
+    fmha_args.nhead_stride_lse  = shape_seqlen_q;
+    fmha_args.nhead_stride_o    = o_perm ? shape_seqlen_q * hdim_v : hdim_v;
+    fmha_args.window_size_left  = 0;
+    fmha_args.window_size_right = 0;
+    fmha_args.mask_type         = static_cast<ck_tile::index_t>(fmha_traits.mask_type);
+
+    fmha_args.s_randval = false;
+    // Since we aren't storing the random values these will be unused for now.
+    fmha_args.stride_randval       = max_seqlen_k;
+    fmha_args.nhead_stride_randval = (shape_seqlen_q * max_seqlen_k);
+    fmha_args.batch_stride_randval = (nhead * shape_seqlen_q * max_seqlen_k);
+
+    result.invoker_factory = [=](const std::vector<Kernel>&) {
+        return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+            decltype(auto) params = raw_params.CastTo<miopen::mha::InvokeParams>();
+            const auto& dataFwd   = params.GetDataForward();
+
+            fmha_fwd_traits fmha_runtime_traits = fmha_traits;
+            fmha_fwd_args fmha_runtime_args     = fmha_args;
+
+            fmha_runtime_args.q_ptr        = dataFwd.qData;
+            fmha_runtime_args.k_ptr        = dataFwd.kData;
+            fmha_runtime_args.v_ptr        = dataFwd.vData;
+            fmha_runtime_args.rand_val_ptr = nullptr;
+            fmha_runtime_args.o_ptr        = dataFwd.oData;
+
+            fmha_runtime_traits.bias_type =
+                dataFwd.biasData != nullptr ? bias_enum::elementwise_bias : bias_enum::no_bias;
+            fmha_runtime_args.bias_ptr = dataFwd.biasData;
+
+            // Top-left causal mask
+            if(dataFwd.mask == miopenMhaMask_t::miopenMhaMaskCausal)
+            {
+                fmha_runtime_traits.mask_type = mask_enum::mask_top_left;
+                fmha_runtime_args.mask_type =
+                    static_cast<ck_tile::index_t>(mask_enum::mask_top_left);
+                fmha_runtime_args.window_size_left  = -1;
+                fmha_runtime_args.window_size_right = 0;
+            }
+
+            fmha_runtime_traits.has_dropout = dataFwd.dropoutProbabilityData != nullptr;
+            float probability               = 0;
+            uint64_t seed                   = 0;
+            uint64_t offset                 = 0;
+            if(fmha_runtime_traits.has_dropout)
+            {
+                hipMemcpy(&probability,
+                          dataFwd.dropoutProbabilityData,
+                          sizeof(float),
+                          hipMemcpyKind::hipMemcpyDeviceToHost);
+                hipMemcpy(&seed,
+                          dataFwd.dropoutSeedData,
+                          sizeof(uint64_t),
+                          hipMemcpyKind::hipMemcpyDeviceToHost);
+                hipMemcpy(&offset,
+                          dataFwd.dropoutOffsetData,
+                          sizeof(uint64_t),
+                          hipMemcpyKind::hipMemcpyDeviceToHost);
+            }
```

Comment:
We have this because there is an API difference between MIOpen and CK.
We are taking in 3 single value tensors for probability, seed, and offset, but CK uses CPU values for their API.

For this current version this is how it needs to work, but I think we need to discuss aligning these to remove the need for this readback.

Reply:
That's kind of usual practice, rocblas does the same for scaling - it allows gpu pointer, but actually copies it back and pass as a kernel parameter.